### ISSUE TITLE
BatchNorm adaptation results

### ISSUE THREAD

**Issue by vanyalzr** [openvinotoolkit/nncf_pytorch#46]
Some additional results regarding #41.

Forgetting (5 batches with momentum = 0.9, then 10 batches with momentum = 0.1) works as well as original resetting to zero and using 200 batches. 

Iterative update of BN statistics layer by layer does not give an accuracy boost. Statistics from previous layers for a given layer are updated on-the-go due to rolling stats calculation and that is sufficient to get good accuracy.

| Model        | Pruning algo info |  Accuracy@1 | Accuracy@5 |
| ------------- |:-------------:|:-------------:|:-------------:|
| ResNet18 (BN adapted original, 200 steps) | geometric median criterion, pruning target = 30% | 33.582 | 59.336 | 
| ResNet18 (BN adapted w/ forgetting, 10 steps) | geometric median criterion, pruning target = 30% | 33.976 | 58.908 | 
| ResNet18 (BN adapted iteratively, 20 steps for each BN node) | geometric median criterion, pruning target = 30% | 33.830 | 58.712 | 

| Model        | Quantization bitwidths | Quantization mode | Range initializer |  Accuracy@1 | Accuracy@5 |
| ------------- |:-------------:|:-------------:|:-------------:|:-------------:|:-------------:|
| ResNet18 (BN adapted original, 200 steps)      |  a8w4 | asymmetric, per-channel for weights | mean min max, 100 batches | 66.866 | 87.476 | 
| ResNet18 (BN adapted w/ forgetting, 10 steps)      |  a8w4 | asymmetric, per-channel for weights | mean min max, 100 batches |  66.798 | 87.490 | 
| ResNet18 (BN adapted iteratively, 20 steps for each BN node)      |  a8w4 | asymmetric, per-channel for weights | mean min max, 100 batches |  66.832 | 87.480 | 
| MobilenetV2 (BN adapted original, 200 steps) | a8w4 |  asymmetric, per-channel for weights | mean min max, 100 batches | 65.216 | 86.304 | 
| MobilenetV2 (BN adapted w/ forgetting, 10 steps) | a8w4 |  asymmetric, per-channel for weights | mean min max, 100 batches | 65.112 | 86.170 | 
| MobilenetV2 (BN adapted iteratively, 20 steps for each BN node) | a8w4 |  asymmetric, per-channel for weights | mean min max, 100 batches | 65.026 | 86.292 |

**Comment by mkaglins** [comment]
**Results on resnet18 of correlation between fine-tuning accuracy and inference accuracy with and without BN-adaptation:**

![vanilla](https://user-images.githubusercontent.com/66625868/86352671-d5f13280-bc6e-11ea-8884-bd17b0ad11e2.png)

![bn_adapted](https://user-images.githubusercontent.com/66625868/86352750-f4efc480-bc6e-11ea-865d-084577cf0cce.png)

![\rho](https://render.githubusercontent.com/render/math?math=\rho&mode=inline) - Pearson correlation coefficient
![\phi](https://render.githubusercontent.com/render/math?math=\phi&mode=inline) - correlation metric proposed in FNNP paper

**Experiment settings:**
- resnet18
- CIFAR100
- fine-tuning for 30 epochs

**Comment by vanyalzr** [comment]
> **Results on resnet18 of correlation between fine-tuning accuracy and inference accuracy with and without BN-adaptation:**
> 
> 
> 
> ![\rho](https://render.githubusercontent.com/render/math?math=%5Crho&mode=inline) - Pearson correlation coefficient
> ![\phi](https://render.githubusercontent.com/render/math?math=%5Cphi&mode=inline) - correlation metric proposed in FNNP paper
> 
> **Experiment settings:**
> 
> * resnet18
> * CIFAR100
> * fine-tuning for 30 epochs

Let's hope results look better for ImageNet and just break in case of CIFAR100.

**Comment by mkaglins** [comment]
**Results on resnet18 of correlation between fine-tuning accuracy and inference accuracy with and without BN-adaptation:**
**(in case of one pruning rate for all layers)**

![vanilla_2](https://user-images.githubusercontent.com/66625868/86518269-0be21280-be39-11ea-9942-0c073ae0e581.png)
![bn_adapted_2](https://user-images.githubusercontent.com/66625868/86518274-13a1b700-be39-11ea-93a5-83434b50f9ae.png)

Graphs show a strong correlation between BN-adapted accuracy and fine-tuned accuracy.

**Experiment settings:**

- resnet18
- CIFAR100
- fine-tuning for 30 epochs
- Pruned with one pruning rate for all layers randomly sampled from (0.0, 0.9]

**Comment by vanyalzr** [comment]
![BNA_mbnetv2_CIFAR100](https://user-images.githubusercontent.com/7557726/86519158-cb869280-be40-11ea-8c4a-d1dfdd71e0ea.png)

| Correlation metric        | Top@1 acc. vanilla vs. fine-tuned           | Top@1 acc. BN adapted vs. fine-tuned  |
| ------------- |:-------------:|:-----:|
| Spearman corr. coef.     | 0.8810 | 0.8415 |
| Weighted Kendall tau      | 0.7347 | 0.7399 |

Details:
- MobilenetV2 fine-tuned on CIFAR100, 65.65 top@1 acc. for original model
- Asymmetic quantization, per-channel for weights, mixed precision confs. (4/8 bits) generated by Bernoulli distribution with 0.35, 0.5, 0.75 expectations for 8 bits
- Fine-tuned for 30 epochs

**Comment by vshampor** [comment]
Closing since BN is already in the code base.
